File: tts.js
const { MsEdgeTTS, OUTPUT_FORMAT } = require('msedge-tts');
const path = require('path');
const fs = require('fs');
const os = require('os');

// Mirrors the same voice map that was in tts.py
const VOICE_MAP = {
    'angelo ph':   'fil-PH-AngeloNeural',
    'danica ph':   'fil-PH-DanicaNeural',
    'james ph':    'fil-PH-JamesNeural',
    'raphael ph':  'fil-PH-RaphaelNeural',
    'aria us':     'en-US-AriaNeural',
    'guy us':      'en-US-GuyNeural',
    'alloy us':    'en-US-AlloyNeural',
    'nanami jp':   'ja-JP-NanamiNeural',
    'keita jp':    'ja-JP-KeitaNeural',
    'haruka jp':   'ja-JP-HarukaNeural',
    'mai jp':      'ja-JP-MaiNeural',
    'ichiro jp':   'ja-JP-IchiroNeural',
    'japanese jp': 'ja-JP-NanamiNeural',
    'jp':          'ja-JP-NanamiNeural',
};

const DEFAULT_VOICE = 'fil-PH-AngeloNeural';

/**
 * Generate a TTS MP3 file using Microsoft Edge TTS (no Python required).
 * @param {string} text       - Text to speak.
 * @param {string|null} voice - Friendly voice name (e.g. "angelo ph"). Falls back to default.
 * @param {string} rate       - Speaking rate as SSML prosody value (e.g. "-10%").
 * @returns {Promise<string>} - Resolves with the absolute path to the generated MP3.
 */
async function generateTTS(text, voice = null, rate = '-10%', guildId = 'global') {
    const voiceKey = (voice || '').toLowerCase().trim();
    const voiceName = VOICE_MAP[voiceKey] || DEFAULT_VOICE;

    const fileName = `tts_${guildId}_${Date.now()}_${Math.floor(Math.random() * 1000)}.mp3`;
    const filePath = path.join(__dirname, fileName);

    const tts = new MsEdgeTTS();
    await tts.setMetadata(voiceName, OUTPUT_FORMAT.AUDIO_24KHZ_96KBITRATE_MONO_MP3);

    const { audioStream } = await tts.toStream(escapeXml(text), { rate });

    await new Promise((resolve, reject) => {
        const writable = fs.createWriteStream(filePath);
        audioStream.pipe(writable);
        writable.on('finish', resolve);
        writable.on('error', reject);
        audioStream.on('error', reject);
    });

    return filePath;
}

function escapeXml(str) {
    return str
        .replace(/&/g, '&amp;')
        .replace(/</g, '&lt;')
        .replace(/>/g, '&gt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&apos;');
}

module.exports = generateTTS;

File: tts.py
import edge_tts
import sys
import asyncio

# Get text from Node.js argument
text = sys.argv[1] if len(sys.argv) > 1 else ""
# Voice selection map: friendly names -> edge-tts voice IDs
VOICE_MAP = {
    "angelo ph": "fil-PH-AngeloNeural",
    "danica ph": "fil-PH-DanicaNeural",
    "james ph": "fil-PH-JamesNeural",
    "raphael ph": "fil-PH-RaphaelNeural",
    "aria us": "en-US-AriaNeural",
    "guy us": "en-US-GuyNeural",
    "alloy us": "en-US-AlloyNeural",
    "nanami jp": "ja-JP-NanamiNeural",
    "keita jp": "ja-JP-KeitaNeural",
    "haruka jp": "ja-JP-HarukaNeural",
    "mai jp": "ja-JP-MaiNeural",
    "ichiro jp": "ja-JP-IchiroNeural",
    "japanese jp": "ja-JP-NanamiNeural",
    "jp": "ja-JP-NanamiNeural"
}

# Voice selection logic
voice_arg = sys.argv[3].lower() if len(sys.argv) > 3 else "angelo ph"
if voice_arg in VOICE_MAP:
    voice = VOICE_MAP[voice_arg]
elif voice_arg in VOICE_MAP.values(): # Allow direct edge-tts IDs
    voice = voice_arg
else:
    voice = VOICE_MAP['angelo ph'] # Default voice if not found

# Output file path from Node.js argument
output = sys.argv[2] if len(sys.argv) > 2 else "tts.mp3"

# Speaking rate (speed) argument from Node.js
rate = sys.argv[4] if len(sys.argv) > 4 else "+0%"

async def main():
    communicate = edge_tts.Communicate(text, voice, rate=rate)
    await communicate.save(output)

# Run async function
asyncio.run(main())